src/selfmain.c

```diff
@@ -14,63 +13,76 @@ extern const uint16_t bootloader_crcs[];
+#define exec_cmdaddr(cmd, addr)                                                                    \
+    do {                                                                                           \
+        NVMCTRL->ADDR.reg = (uint32_t)addr;                                                        \
+        NVMCTRL->CTRLB.reg = NVMCTRL_CTRLB_CMDEX_KEY | cmd;                                        \
+        while (NVMCTRL->STATUS.bit.READY == 0) {                                                   \
+        }                                                                                          \
     } while (0)
 #endif
+#define exec_cmd(cmd) exec_cmdaddr(cmd, NVMCTRL_USER)
 
 void setBootProt(int v) {
-    uint32_t fuses[2];
-
-    #ifdef SAMD21
-    while (!(NVMCTRL->INTFLAG.reg & NVMCTRL_INTFLAG_READY)) {}
-    #endif
-    #ifdef SAMD51
-    while (NVMCTRL->STATUS.bit.READY == 0) {}
+    #if defined(SAMD21)
+        uint32_t fuses[2], 
+                 newfuses[2];
+        while (!(NVMCTRL->INTFLAG.reg & NVMCTRL_INTFLAG_READY)) {
+        }
+    #elif defined(SAMD51)
+        uint32_t fuses[128],    // 512 bytes (whole user page)
+                 newfuses[5];
+        while (!NVMCTRL->STATUS.bit.READY) {
+        }
     #endif
 
-    fuses[0] = *((uint32_t *)NVM_FUSE_ADDR);
-    fuses[1] = *(((uint32_t *)NVM_FUSE_ADDR) + 1);
+    memcpy(fuses, (uint32_t *)NVM_FUSE_ADDR, sizeof(fuses));
+    memcpy(newfuses, fuses, sizeof(newfuses)); // Start with new values equal to current
 
-    bool repair_fuses = false;
     // Check for damaged fuses. If the NVM user page was accidentally erased, there
     // will be 1's in wrong places. This would enable the watchdog timer and cause other
     // problems. So check for all ones outside of the SAMD21/51 BOOTPROT fields, or all ones
     // in fuses[1]. If it appears the fuses page was erased, replace fuses with reasonable values.
-    if ((fuses[0] & 0x03fffff0) == 0x03fffff0 || fuses[1] == 0xffffffff) {
-        repair_fuses = true;
-
-        // These canonical fuse values taken from working Adafruit SAMD21 and SAMD51 boards.
-        #ifdef SAMD21
-        fuses[0] = 0xD8E0C7FA;
-        fuses[1] = 0xFFFFFC5D;
-        #endif
-        #ifdef SAMD51
-        fuses[0] = 0xF69A9239;
-        fuses[1] = 0xAEECFF80;
+    bool repair_fuses = ((fuses[0] & 0x03fffff0) == 0x03fffff0 || fuses[1] == 0xffffffff);
+    #if defined(SAMD51)
+        // SAMD51 needs an additional check
+        repair_fuses |= (fuses[4] == 0xffffffff);
+    #endif
+    if (repair_fuses) {
+        // These canonical fuse values taken from working Adafruit boards.
+        #if defined(SAMD21)
+            newfuses[0] = 0xD8E0C7FF;
+            newfuses[1] = 0xFFFFFC5D;
+        #elif defined(SAMD51)
+            newfuses[0] = 0xFE9A9239;
+            newfuses[1] = 0xAEECFF80;
+            newfuses[2] = 0xFFFFFFFF;
+            newfuses[3] = 0xFFFFFFFF;
```

Comment:
`newfuses[3]` This is a user-written value, in between two fuse or factory values. So you can fetch this from the existing value. Don't overwrite.

Reply:
I am here when data is erased. I don't have "existing value"

Reply:
but since it's user fuse, correct to not change it